source/Nevermore/Advanced/ReadTransaction.cs

```diff
@@ -521,8 +531,9 @@ IEnumerable<TRecord> Execute()
                     yield return item;
             }
 
-            return configuration.SupportConcurrentExecution
-                ? new ThreadSafeEnumerable<TRecord>(Execute, DeadlockAwareLock)
+            // Use the thread safe variant if we know it's going to be used
+            return configuration.ConcurrencyMode is not ConcurrencyMode.NoLock
```

Comment:
Suggestion: A clearer and more reliable way to make this decision could be

- Add a `bool IsThreadSafe { get; }` property to `TransactionConcurrencyHandler`
- The implementations can return true/false appropriately

`return TransactionConcurrencyHandler.IsThreadSafe ? new ThreadSafeEnumerable<TRecord>...`

Reply:
Underlying principle here: You've got one input - the `ConcurrencyMode` enum, but you make decisions about it in two places. If the decisions don't relate to each other, that's fine -- however in this case the choice to use ThreadSafeEnumerable, and the behaviour of the `TransactionConcurrencyHandler` are _highly_ related.

Technique: Reduce the decision-points (and thus, risk of bugs) by having one decision "flow through" to the next. If that all sounds like gibberish, hit me up for a zoom tomorrow :smile:

Reply:
I'm keen to follow up here on zoom. Another concern here is that we still want to use the logging-only concurrency handler.
I think the cleanest way forward is to always return the `ThreadSafeEnumerable`, and give it what ever concurrency handler we've got. The only thing I don't like about that is the fact that we can return a `ThreadSafeEnumerable` with a concurrency handler that doesn't provide thread safety, but we need it for the logging anyway 😕